runtime/jsMain/src/kotlinx/serialization/DynamicObjectSerializer.kt

```diff
@@ -0,0 +1,130 @@
+        serializer.encode(strategy, obj)
+        return serializer.result
+    }
+
+    @ImplicitReflectionSerializer
+    inline fun <reified T : Any> serialize(obj: List<T?>): dynamic {
+
+        val serializer = DynamicObjectEncoder()
+        val strategy = context.getContextualOrDefault(T::class).nullable.list
+        serializer.encode(strategy, obj)
+
+        return serializer.result
+    }
+}
+
+class DynamicObjectEncoder : ElementValueEncoder() {
+    private object NoOutputMark
+
+    private val json: Json = Json(JsonConfiguration.Stable)
+
+    var result: dynamic = NoOutputMark
+
+    internal data class Node(val mode: WriteMode, val jsObject: dynamic) {
+        var index: Int = 0
+        lateinit var parent: Node
+    }
+
+    private lateinit var current: Node
+
+    private var writingFirst = true
+    var currentName: String = ""
+
+
+    override fun encodeValue(value: Any) {
+        if (current.mode == WriteMode.MAP && current.index % 2 == 0) {
+            require(value is String || value is Int) { "only String or Int keys in maps are supported" }
+            currentName = value.toString()
+
+        } else {
+            current.jsObject[currentName] = value
+        }
+
+    }
+
+    override fun encodeChar(value: Char) {
+        current.jsObject[currentName] = value.toString()
+    }
+
+    override fun encodeNull() { // encode as undefined
+        // TODO when do we omit a nullable property and when do we write null explicitly?
+    }
+
+    private fun indent(jsObject: dynamic, mode: WriteMode) {
+        val next = Node(mode, jsObject)
+        next.parent = current
+        current = next
+    }
+
+    private fun unindent() {
+        current = current.parent
+    }
+
+    override fun encodeElement(desc: SerialDescriptor, index: Int): Boolean {
+        current.index = index
+
+        if (current.mode != WriteMode.MAP) {
+            // TODO always legal javascript name? are we sure?
+            // TODO  name is always the index in a list, are we allowed to (ab)use that?
+            currentName = desc.getElementName(index)
+        }
+        return true
+    }
+
+    override fun beginStructure(desc: SerialDescriptor, vararg typeParams: KSerializer<*>): CompositeEncoder {
+        val newMode = json.switchMode(desc)
+        if (writingFirst) {
+            writingFirst = false
+
+            result = newStructure(newMode)
+            current = Node(newMode, result)
+            current.parent = current
+        } else {
+
+            val next = newStructure(newMode)
+            current.jsObject[currentName] = next
+            indent(next, newMode)
+        }
+
+        current.index = 0
+        return this
+    }
+
+    internal fun newStructure(mode: WriteMode) = when (mode) {
+        WriteMode.OBJ -> js(BEGIN_OBJ.toString() + END_OBJ)
+        WriteMode.MAP -> js(BEGIN_OBJ.toString() + END_OBJ)
+        WriteMode.LIST -> js(BEGIN_LIST.toString() + END_LIST)
+        WriteMode.POLY_OBJ -> TODO()
```

Comment:
The reason we don't just use `begin`/`end` from the `WriteMode` itself (and essentially redeclare them here) is because the compiler doesn't treat `begin`/`end` as string constants, correct?